parsing/parchmint-parser.js

```diff
@@ -254,15 +320,95 @@ class ParchmintParser {
     }
 
     /**
-     * Parse a single Port object from the given JSON object.
+     * Parse a Terminal object from the given JSON object.
+     *
+     * The component is only searched for on the given layer, so as to get the
+     * correct Port object.
      *
      * @since 1.0.0
      *
-     * @param {object}  portObj A JSON object with fields layer, label, x, and y.
-     * @returns {Port}  The resulting Port object.
+     * @param {object}  termObj A JSON object with fields "component" and
+     *                          "port".
+     * @param {string}  layer   The ID of the Layer on which to search for the
+     *                          Component.
+     * @returns {Terminal}  A Terminal object with the Component and Port if
+     *                      they could both be found, or a default Terminal
+     *                      object otherwise.
      */
-    static parsePort(portObj) {
-        return new Port(portObj['label'], this.parseCoord(portObj));
+    parseTerminal(termObj, layer) {
+        let comp = null;
+
+        // First let's find the Component we need
+        this.components.get(layer).forEach(value => {
+            if (value.id === termObj['component']) {
+                comp = value;
+            }
+        });
+
+        if (comp) {
+            // If we found a component lets get the Port we need from it
+            for (let value of comp.ports) {
+                if (value.label === termObj['port']) {
+                    return new Terminal(comp, value);
+                }
+            }
+        } else {
+            // We didn't find a component which means the Parchmint is invalid
+            this.valid = false;
+            console.log('Parser: Unable to find Component with ID "' + termObj['component'] + '" on Layer with ID "' +
+                    layer + '". Parsing will continue, using a default Terminal object.');
+            return new Terminal();
+        }
+
+        // We didn't find the a Port so the Parchmint is invalid, but at the very least we can put the Component in.
+        this.valid = false;
+        console.log('Parser: Unable to find Port with label "' + termObj['port'] + '" on Component with ID "' +
+                termObj['component'] + '". Parsing will continue using a Terminal object with a null Port.');
+        return new Terminal(comp);
```

Comment:
You have an if/else above so I don't think this code will ever get called.

Reply:
The if statement only returns if a port is found. If not then we return a Terminal with only the Component reference. We go into the else if we couldn't find the Component and return a default Terminal object.

Reply:
Oh I think I see, so if the `for` loop in the `if` statement doesn't match then we will fall out of that and bypass the `else` to run the last section, is that correct?

Reply:
Exactly